### PR TITLE
Support Custom MaxText model (with vLLM engine) in RL rollouts.

### PR DESCRIPTION
**Pull request by NicoGrande**
# Description

*This PR finishes the work started by @gagika in #2767. Credits to @gagika for helping with this feature!*

This PR adds the changes required to `train_rl.py` as well as other modules related to Tunix integration to allow for additional configurations needed for the MaxText on vLLM flow to be passed to Tunix. 

More specifically, this PR adds `vllm_additional_config` and  `vllm_hf_config_path` as new arguments such that these values can be pipelined to Tunix for RL. 

Additionally, this PR makes some small modifications to `tunix_adapter.py` to allow for no-ops to be used as mappings when running RL using MaxText for vLLM. 

# Tests

## Gemma3-4B:

**Local (v6e-4 VM):**
```
NEW_MODEL_DESIGN=True  HF_TOKEN=$HF_TOKEN TPU_BACKEND_TYPE=jax python3 -m src.MaxText.rl.train_rl src/MaxText/configs/rl.yml \
  model_name=gemma3-4b \
  tokenizer_path=google/gemma-3-4b-it \
  run_name=$WORKLOAD \
  base_output_directory=$OUTPUT_PATH \
  hf_access_token=$HF_TOKEN \
scan_layers=False \
load_parameters_path="gs://maxtext-gemma/unified/gemma3/4b/unscanned/2025-08-09-01-17/0/items" \
  vllm_hf_config_path=src/MaxText/integration/vllm/maxtext_vllm_adapter   vllm_additional_config='{"maxtext_config": {"model_name": "gemma3-4b", "max_prefill_predict_length": 28, "max_target_length": 32, "ici_tensor_parallelism": 4}}'
```

Output: [logs](https://paste.googleplex.com/6325348478353408)

## Qwen3-30B-A3B:

**v5p-64 Cluster:**
```
xpk workload create-pathways --workload $WORKLOAD_NAME --num-slices=1 --priority very-high --docker-image gcr.io/tpu-prod-env-multipod/maxtext-vllm-tpu-test:stable --cluster $CLUSTER_NAME --tpu-type=$TPU_TYPE --zone=$ZONE --project=$PROJECT_ID --command "cd /app && pip install --no-deps -e . && VLLM_TORCH_PROFILER_DIR=$OUTPUT_PATH MAXTEXT_PKG_DIR=/app/src/MaxText HF_TOKEN=$HF_TOKEN JAX_RANDOM_WEIGHTS=1 VLLM_ENABLE_V1_MULTIPROCESSING=0 NEW_MODEL_DESIGN=1 TPU_MIN_LOG_LEVEL=0 TF_CPP_MIN_LOG_LEVEL=0 TPU_STDERR_LOG_LEVEL=0 JAX_PLATFORMS=proxy,cpu JAX_BACKEND_TARGET=grpc://127.0.0.1:29000 ENABLE_PATHWAYS_PERSISTENCE=1 python3 -m src.MaxText.rl.train_rl src/MaxText/configs/rl.yml \
  model_name=qwen3-30b-a3b \
  tokenizer_path=Qwen/Qwen3-30B-A3B \
  run_name=$WORKLOAD \
  base_output_directory=$OUTPUT_PATH \
  hf_access_token=$HF_TOKEN \
  batch_size=16 \
  rollout_data_parallelism=4 \
  rollout_tensor_parallelism=4 \
  hbm_utilization_vllm=0.60 \
  scan_layers=False \
  load_parameters_path=$CHECKPOINT_PATH \
  allow_split_physical_axes=True \
  vllm_hf_config_path=src/MaxText/integration/vllm/maxtext_vllm_adapter vllm_additional_config='{maxtext_config: {model_name: qwen3-30b-a3b, allow_split_physical_axes: true, log_config: false}}'"
```
Output: [logs](https://cloudlogging.app.goo.gl/KZK3eYpLZnNnRHMh8)

# Checklist

Before submitting this PR, please make sure (put X in square brackets):
- [x] I have performed a self-review of my code. For an optional AI review, add the `gemini-review` label.
- [x] I have necessary comments in my code, particularly in hard-to-understand areas.
- [x] I have run end-to-end tests tests and provided workload links above if applicable.
- [x] I have made or will make corresponding changes to the doc if needed, including adding new documentation pages to the relevant Table of Contents (toctree directive) as explained in [our documentation](https://maxtext.readthedocs.io/en/latest/development.html#adding-new-documentation-files).

### CODE REVIEW

File: src/MaxText/configs/vllm.yml [hunk 1]
```diff
@@ -49,13 +50,23 @@ logical_axis_rules: [
...
+                      ['tensor', ['model']],
+                      ['tensor_transpose', ['model']],
+                      ['fsdp', ['data']],
+                      ['fsdp_transpose', ['data']],
+                      ['sequence', ['model']],
+                      ['context', ['model']],
                     ]
 data_sharding: [['data', 'model', 'expert']]
-input_data_sharding_logical_axes: ['activation_embed_and_logits_batch']
+input_data_sharding_logical_axes: ['activation_embed_and_logits_batch']
```
Comment (A9isha) [inline src/MaxText/configs/vllm.yml]: nit: please add the new line

File: src/MaxText/configs/vllm.yml [hunk 2]
```diff
@@ -49,13 +50,23 @@ logical_axis_rules: [
                       ['heads', ['model']],
                       ['q_heads', ['model']],
                       ['kv_heads', ['model']],
+                      ['kv_head_dim', []],
```
Comment (A9isha) [inline src/MaxText/configs/vllm.yml:55]: for my understanding: why isn't `kv_head_dim` and `kv` not to be ever sharded?

File: src/MaxText/configs/vllm.yml [hunk 3]
```diff
@@ -49,13 +50,23 @@ logical_axis_rules: [
...
                       ['cache_heads', ['model']],
                       ['exp', ['expert']],
                       ['paged_kv_heads', ['model']],
+                      ['autoregressive', ['model']],
+                      ['tensor', ['model']],
+                      ['tensor_transpose', ['model']],
+                      ['fsdp', ['data']],
+                      ['fsdp_transpose', ['data']],
+                      ['sequence', ['model']],
+                      ['context', ['model']],
```
Comment (A9isha) [inline src/MaxText/configs/vllm.yml]: curious why is `context/sequence` being mapped to `model` aka `TP` and `fsdp` to DP?

File: src/MaxText/integration/vllm/maxtext_vllm_adapter/adapter.py [hunk 4]
```diff
@@ -199,9 +184,14 @@ def load_weights(self, rng_key: jax.Array) -> None:
     Args:
       rng_key: A JAX random key for model initialization.
     """
-    self.model, _ = model_creation_utils.create_nnx_model(
-        self.maxtext_config, mesh=self.mesh, model_mode=self.model_mode, rng_key=rng_key
-    )
+    if self.model is not None:
+      return
+
+    with nn.logical_axis_rules(""):
```
Comment (A9isha) [inline src/MaxText/integration/vllm/maxtext_vllm_adapter/adapter.py:190]: curious, what happens when we don't specify `with nn.logical_axis_rules(""):`

File: src/MaxText/rl/evaluate_rl.py [hunk 5]
```diff
@@ -121,13 +121,18 @@ def score_responses(tmvp_config, question, responses, answer):
 
     # Check exact correctness
     try:
-      if float(extracted_response.strip()) == float(answer.strip()):
+      # Remove ',' and '$' then convert to float
+      val_extracted = float(extracted_response.replace(",", "").replace("$", "").strip())
```
Comment (A9isha) [inline src/MaxText/rl/evaluate_rl.py:125]: Should we be removing `replace("$", "")`?

File: src/MaxText/rl/evaluate_rl.py [hunk 6]
```diff
@@ -121,13 +121,18 @@ def score_responses(tmvp_config, question, responses, answer):
...
+      val_answer = float(answer.replace(",", "").replace("$", "").strip())
+
+      if val_extracted == val_answer:
         is_correct = True
 
       # Check partial correctness (within 10%)
-      ratio = float(extracted_response.strip()) / float(answer.strip())
-      if 0.9 <= ratio <= 1.1:
-        is_partially_correct = True
+      if val_answer != 0.0:
```
Comment (A9isha) [inline src/MaxText/rl/evaluate_rl.py]: why are we introducing this condition?

File: src/MaxText/rl/train_rl.py [hunk 7]
```diff
@@ -423,7 +444,12 @@ def rl_train(trainer_config, sampler_config, trainer_devices, sampler_devices):
       max_logging.log(
           "enable_tunix_perf_metrics is True but tunix.perf modules are not available, skipping Tunix-managed metrics."
       )
-  with nn_partitioning.axis_rules(trainer_config.logical_axis_rules):
+
+  vllm_config_path = epath.Path(MAXTEXT_PKG_DIR) / "configs" / "vllm.yml"
```
Comment (A9isha) [inline src/MaxText/rl/train_rl.py:448]: The rest of the MaxText repo contains `os.path.join`. Should we make this consistent?

File: src/MaxText/configs/vllm.yml (hunk 1 above)
Comment (NicoGrande) [reply src/MaxText/configs/vllm.yml]: Thanks for pointing out! Added.

File: src/MaxText/configs/vllm.yml (hunk 2 above)
Comment (NicoGrande) [reply src/MaxText/configs/vllm.yml:55]: I'm not 100% sure about this one, perhaps @richjames0 can speak more to this, but these are also not sharded in base.yml. If we should be sharding these I am happy to modify them.

File: src/MaxText/configs/vllm.yml (hunk 3 above)
Comment (NicoGrande) [reply src/MaxText/configs/vllm.yml]: context, sequence and fsdp do not exist in the vLLM mesh. So we need to provide alternatives for these types of parallelisms. I think it may make more sense to map sequence and context to data in case of vLLM. Thoughts @richjames0 ?

File: src/MaxText/integration/vllm/maxtext_vllm_adapter/adapter.py (hunk 4 above)
Comment (NicoGrande) [reply src/MaxText/integration/vllm/maxtext_vllm_adapter/adapter.py:190]: It messes up the model sharding at initialization - we were seeing the logical axis rules set to `None` if we didn't have this context manager in place. This was causing us to replicate the model unexpectedly and OOM when testing vLLM.

File: src/MaxText/rl/evaluate_rl.py (hunk 5 above)
Comment (NicoGrande) [reply src/MaxText/rl/evaluate_rl.py:125]: Could you elaborate on why? I think this line is in place to handle evals which involve dollar amounts.

File: src/MaxText/rl/evaluate_rl.py (hunk 6 above)
Comment (NicoGrande) [reply src/MaxText/rl/evaluate_rl.py]: Good catch - I think this should always be run.

File: src/MaxText/rl/train_rl.py (hunk 7 above)
Comment (NicoGrande) [reply src/MaxText/rl/train_rl.py:448]: I saw `epath` in a couple of other files including `train_rl.py`. I would suggest we keep this since it handles both local files and GCS with the same API :) 

If you are referring to the `/ "configs" / "vllm.yml"` portion of it, I find this approach to be a bit more readable and usually how `pathlib` is used. Since `epath` is basically a `pathlib` wrapper, I went with this approach but I'm happy to change it if we have a strong stance on this.

File: src/MaxText/configs/vllm.yml (hunk 2 above)
Comment (richjames0) [reply src/MaxText/configs/vllm.yml:55]: @gobbleturk is the expert here but I believe the TLDR is that it's so small the comms kill you

File: src/MaxText/configs/vllm.yml (hunk 3 above)
Comment (richjames0) [reply src/MaxText/configs/vllm.yml]: Discussed with Nico and probably we don't need/want these axes. Thanks for spotting @A9isha !

Comment (NicoGrande) [reply src/MaxText/configs/vllm.yml]: Synced offline with @richjames0 and confirmed with a quick experiment that these are not needed.

File: src/MaxText/rl/train_rl.py (hunk 7 above)
Comment (NicoGrande) [reply src/MaxText/rl/train_rl.py:448]: Changed to os.path.join.

File: src/MaxText/rl/evaluate_rl.py (hunk 5 above)
Comment (NicoGrande) [reply src/MaxText/rl/evaluate_rl.py:125]: Synced offline.
